src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/SumNormalization.java

```diff
@@ -0,0 +1,169 @@
+			}
+			final BasicAtom aggregateTypeAtom = (BasicAtom) makePredicatesInternal(parse(
+				"aggregate_function(aggregate(" + ConstantTerm.getInstance(aggregateCount) + "), \"#sum\").")).getFacts().get(0);
+			additionalFacts.add(aggregateTypeAtom);
+			lastAggregateCount = aggregateCount;
+		}
+		// Leave program as-is if no aggregates occur.
+		if (additionalRules.isEmpty()) {
+			return;
+		}
+		Program summationEncoding = makePredicatesInternal(new ProgramParser().parse(summationSubprogram));
+		summationEncoding.accumulate(makePredicatesInternal(new Program(Collections.emptyList(),
+			additionalFacts, new InlineDirectives())));	// Add internalised type facts.
+		summationEncoding.getRules().addAll(additionalRules);
+
+		// Add enumeration rule that uses the special EnumerationAtom.
+		// The enumeration rule is: "input_number_with_first(A, I, F) :- input_with_first(A, X, F), _index(A, X, I)."
+		Rule enumerationRule = makePredicatesInternal(parse("input_number_with_first(A, I, F) :- input_with_first(A, X, F).")).getRules().get(0);
+		EnumerationAtom enumerationAtom = new EnumerationAtom(parse("index(A, X, I).").getFacts().get(0).getTerms());
+		enumerationRule.getBody().add(enumerationAtom.toLiteral());
+		summationEncoding.getRules().add(enumerationRule);
+
+		// Add sum encoding to program.
+		inputProgram.accumulate(summationEncoding);
+	}
+
+	private List<Rule> rewriteAggregates(Rule rule) {
+
+		// Example rewriting/connection:
+		// x :- 6 <= #sum {3,a:a; 4,b:b; 5,c:c}.
+		// is rewritten to:
+		// x :- output(aggregate(1), 6).
+		// input_with_first(aggregate(1), element_tuple(3, a), 3) :- a.
+		// input_with_first(aggregate(1), element_tuple(4, b), 4) :- b.
+		// input_with_first(aggregate(1), element_tuple(5, c), 5) :- c.
+		// bound(aggregate(1), 6).
+
+		// Create interface atoms to the aggregate encoding.
+		final BasicAtom aggregateOutputAtom = (BasicAtom) makePredicatesInternal(parse(
+			"output(aggregate(AGGREGATE_ID), LOWER_BOUND).")).getFacts().get(0);
+		final BasicAtom aggregateInputAtom = (BasicAtom) makePredicatesInternal(parse(
+			"input_with_first(aggregate(AGGREGATE_ID), ELEMENT_TUPLE, FIRST_VARIABLE).")).getFacts().get(0);
+		final BasicAtom lowerBoundAtom = (BasicAtom) makePredicatesInternal(parse(
+			"bound(aggregate(AGGREGATE_ID), LOWER_BOUND).")).getFacts().get(0);
+
+
+
+
+		ArrayList<Literal> aggregateOutputAtoms = new ArrayList<>();
+		int aggregatesInRule = 0;	// Only needed for limited rewriting.
+		ArrayList<Rule> additionalRules = new ArrayList<>();
+
+		for (Iterator<Literal> iterator = rule.getBody().iterator(); iterator.hasNext();) {
+			Literal bodyElement = iterator.next();
+			// Skip non-aggregates.
+			if (!(bodyElement instanceof AggregateLiteral)) {
+				continue;
+			}
+			AggregateLiteral aggregateLiteral = (AggregateLiteral) bodyElement;
+			AggregateAtom aggregateAtom = aggregateLiteral.getAtom();
+
+			// FIXME: limited rewriting of lower-bounded cardinality/sum aggregates only.
```

Comment:
Please either resolve remaining inline TODOs before merging to master or create issues to track them.